### PR TITLE
Rate limit handling for Dropbox

### PR DESCRIPTION
**Pull request by GGGuenni**
Got a bit bigger than anticipated because of refactors
Could not test it as I did not want to rate limit the dropbox right now therefore draft
Also refactored some popup code to reduce repetition

### CODE REVIEW

**Comment by SomeTroglodyte** [comment]
Didn't I code a ratelimit aware mod query? `object RateLimit {` line 128?

**Comment by GGGuenni** [comment]
Ahh I haven't seen you already implemented something for Github
KK will see how to use it for dropbox

**Comment by SomeTroglodyte** [comment]
Will need lots of testing - github's limit is absolutely unfair hard-chunked time, not a sliding window as would have been almost equally easy to implement... But it's only a minute so testing _was_ easy for that one.

**Comment by GGGuenni** [comment]
Looked into it and I don't think bending the GitHub rate limit implementation to support the Dropbox rate limit would be a good idea. 
Dropbox rate limit is simply: get limited and receive time -> wait for time before trying again. 

Just tested during the latest rate limit and works fine

File: core/src/com/unciv/logic/multiplayer/DropBox.kt [hunk 1]
```diff
@@ -56,8 +65,71 @@ object DropBox {
...
+
+    /**
+     * If the dropbox rate limit is reached for this bearer token we strictly have to wait for the
+     * specified retry_after seconds before trying again. If non is supplied or can not be parsed
+     * the default value of 5 minutes will be used.
+     * Any attempt before the rate limit is dropped again will also contribute to the rate limit
+     */
+    private fun triggerRateLimit(error: ErrorResponse) {
+        try {
+            remainingRateLimitSeconds = error.details?.retry_after?.toInt() ?: 300
```
Comment (SomeTroglodyte) [inline core/src/com/unciv/logic/multiplayer/DropBox.kt:118]: Would toIntOrNull make the try-catch redundant?

**Comment by yairm210** [comment]
I think this is ready, I hope you tested it enough that it won't break existing games

**Review by yairm210 — APPROVED** [review]

File: core/src/com/unciv/logic/multiplayer/DropBox.kt [hunk 2]
```diff
@@ -127,42 +165,29 @@ object DropBox {
...
-        return DropBox.downloadFileAsString(getLocalGameLocation(fileName))
-    }
-
-    override fun getFileMetaData(fileName: String): IFileMetaData {
-        return DropBox.getFileMetaData(getLocalGameLocation(fileName))
-    }
+    @Suppress("PropertyName")
+    private class ErrorResponse {
+        var error_summary = ""
+        var details: Details? = null
```
[the remark this answers is not in the document]
Comment (GGGuenni) [reply core/src/com/unciv/logic/multiplayer/DropBox.kt:187]: The docs say so you are right, but I'm pretty sure I've always received a details tag 🤔
Did you check the api response yourself?

**Comment by GGGuenni** [comment]
> that it won't break existing games
I don't see how it could break existing games so should be good to go

Thanks to @Azzurite for finishing up this PR! Really appreciate it!

**Comment by touhidurrr** [comment]
Ok, correct me if I am wrong, but I don’t really like this implementation.
Currently each time we get 429 we throw a RateLimit execption and tell players to wait for retryAfter time. But that's not how Rate Limits are supposed to be handled and is very inefficient.
What we should do is calculate and store a timestamp that holds the time untill Rate Limit is finished and then every time we are supposed to check for game state we would just do if (currentTime < timestamp) break.
That is the correct way to handle rate limits as you are not supposed to make a new network request in tge first place if the server sends you a rateLimit error.

So, I would suggest / request this pr to not be merged unless this argument is considered.

**Comment by touhidurrr** [comment]
Also I think Dropbox has separate rate limit for upload & download. So, if the timestamp method is implemented, we need to take that into consideration too.

**Comment by touhidurrr** [comment]
> > as you are not supposed to make a new network request in tge first place if the server sends you a rateLimit error.
> 
> That's what throwing the exception is for. If it's thrown, no network request is sent. Throwing an exception works kinda like a `return`, returning an exception object. The UI can then handle the exception which contains the remaining seconds for the rate limit however it sees fit. Instead of a timestamp, I went with a countdown which refreshes every second until the rate limit time is over. So it's basically the same. Performance-wise, I don't think there is a huge difference especially since this is nothing that has to be super optimized.
> 
> > Also I think Dropbox has separate rate limit for upload & download. So, if the timestamp method is implemented, we need to take that into consideration too.
> 
> As stated in the [Dropbox Performance Guide](https://developers.dropbox.com/de-de/dbx-performance-guide)
> 
> > The [Dropbox API](https://www.dropbox.com/developers/documentation/http/overview) enforces rate limits on the number of API calls issued over a period of time on a per-authorization (token) basis
> 
> Which is also what I've been observing. If our token is rate limited, every API call, no matter if delete, download, upload, or get_metadata, is blocked.

That's why I hid the comment minutes after I wrote it. Initially I had only seen the exception part that time and thought that was how you were writing this pr. Then wrote a comment, read more, realized that I was wrong and hid the comment.
I kept the second comment anyways cause I wanted to know how Dropbox rate limitng work for multiple operations cause one time when making a download request to the api i got a 429 error. However the following upload request that I made went through. And so I was thought maybe Dropbox has separate rate limits for various operations. Cloudflare KV that I was using back then had similar limits. 100k/day for read and 10k/day for write.

Anyways thanks for explaining things like this. Also I think that showing countdown until rate limiting has finished is a cool thing. XD

**Review by touhidurrr — APPROVED** [review]

**Review by touhidurrr — APPROVED** [review]

**Pull request merged**